pkg/providers/gcp/provider_redis.go

```diff
@@ -36,30 +43,122 @@ func (rp RedisProvider) SupportsStrategy(deploymentStrategy string) bool {
+		return nil, croType.StatusMessage("network ip address range creation in progress"), nil
+	}
+	logger.Infof("created ip address range %s: %s/%d", address.GetName(), address.GetAddress(), address.GetPrefixLength())
+
+	logger.Infof("creating network service connection")
+	service, err := networkManager.CreateNetworkService(ctx)
+	if err != nil {
+		msg := "failed to create network service"
+		return nil, croType.StatusMessage(msg), errorUtil.Wrap(err, msg)
+	}
+	if service == nil {
+		return nil, croType.StatusMessage("network service connection creation in progress"), nil
+	}
+	logger.Infof("created network service connection %s", service.Service)
+
+	//TODO implement me
+	return p.createRedisCluster(ctx, r)
+}
+
+func (p *RedisProvider) createRedisCluster(ctx context.Context, r *v1alpha1.Redis) (*providers.RedisCluster, croType.StatusMessage, error) {
 	// TODO implement me
-	return nil, "", nil
+	return nil, croType.StatusMessage("successfully created and tagged, gcp elasticache"), nil
 }
 
-func (rp RedisProvider) DeleteRedis(ctx context.Context, r *v1alpha1.Redis) (types.StatusMessage, error) {
-	_, err := rp.CredentialManager.ReconcileProviderCredentials(ctx, r.Namespace)
+func (p *RedisProvider) DeleteRedis(ctx context.Context, r *v1alpha1.Redis) (croType.StatusMessage, error) {
+	logger := p.Logger.WithField("action", "DeleteRedis")
+	logger.Infof("reconciling delete redis %s", r.Name)
+
+	creds, err := p.CredentialManager.ReconcileProviderCredentials(ctx, r.Namespace)
 	if err != nil {
 		errMsg := fmt.Sprintf("failed to reconcile gcp redis provider credentials for redis instance %s", r.Name)
-		return types.StatusMessage(errMsg), fmt.Errorf("%s: %w", errMsg, err)
+		return croType.StatusMessage(errMsg), fmt.Errorf("%s: %w", errMsg, err)
 	}
+
 	// TODO implement me
-	return "", nil
+	isLastResource, err := resources.IsLastResource(ctx, p.Client)
+	if err != nil {
+		errMsg := "failed to check if this cr is the last cr of type postgres and redis"
+		return croType.StatusMessage(errMsg), errorUtil.Wrap(err, errMsg)
+	}
+
+	networkManager, err := NewNetworkManager(ctx, option.WithCredentialsJSON(creds.ServiceAccountJson), p.Client, logger)
+	if err != nil {
+		errMsg := "failed to initialise network manager"
+		return croType.StatusMessage(errMsg), errorUtil.Wrap(err, errMsg)
+	}
+
+	return p.deleteRedisCluster(ctx, networkManager, r, isLastResource)
 }
 
-var _ providers.RedisProvider = (*RedisProvider)(nil)
+// TODO
+func (p *RedisProvider) deleteRedisCluster(ctx context.Context, networkManager NetworkManager, r *v1alpha1.Redis, isLastResource bool) (croType.StatusMessage, error) {
+
+	// TODO delete GCP redis cluster
+
+	// remove networking components
+	if isLastResource {
+		if err := networkManager.DeleteNetworkPeering(ctx); err != nil {
+			msg := "failed to delete cluster network peering"
+			return croType.StatusMessage(msg), errorUtil.Wrap(err, msg)
```

Comment:
We seem to be using a mixture of `errorUtil.Wrap(err, msg)` and `fmt.Errorf("%s: %w", msg, err)`.
It would be good to agree on one as it would make the error handling style more consistent. I've used both and don't have a strong opinion on either one.
`errorUtil.Wrap(err, msg)` returns a nil error if the `err` passed to it is nil and ignores the message.
`fmt.Errorf("%s: %w", msg, err)` would always return an error with the message regardless if the `err` passed to it is nil.
Found a discussion about the same issue here: https://github.com/golang/go/issues/38193

Reply:
My go to is:

- `errorUtil.Wrap` - if we have an error
- `fmt.Errorf` - if we have an object/string that needs to be formatted in the error message but no error
- `errors.New` - if we are only outputting a message with no formatted object and no error